lib/src/main/java/org/team1540/rooster/drive/pipeline/AdvancedArcadeJoystickInput.java

```diff
@@ -0,0 +1,95 @@
+package org.team1540.rooster.drive.pipeline;
+
+import java.util.OptionalDouble;
+import java.util.function.DoubleSupplier;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Modified arcade drive joystick input.
+ *
+ * This arcade drive uses one throttle input and two wheel inputs (soft and hard). The soft wheel
+ * input is multiplied by the absolute value of the throttle input before being used, while the hard
+ * is not&mdash;this allows for smoother control of the robot's path at a variety of speeds. The
+ * core algorithm is adapted from Team 2471's drive code, which can be found
+ * <a href="https://github.com/TeamMeanMachine/2018FRC/blob/13c96d2f0e2e780b0cec03fe71ad4919f70f6368/src/main/kotlin/org/team2471/frc/powerup/drivetrain/Drivetrain.kt#L163">here</a>.
+ * <p>
+ * This class is an {@link Input} that provides a {@link TankDriveData}. The resulting {@link TankDriveData} will have the left and right velocities set
+ * with units corresponding to the max velocity set on construction, as well as the turning rate in
+ * radians per second. All other values are empty {@link OptionalDouble OptionalDoubles}.
+ *
+ * @see Input
+ * @see FeedForwardProcessor
+ */
+public class AdvancedArcadeJoystickInput implements Input<TankDriveData> {
+
+  private double maxVelocity;
+  private double trackWidth;
+  private @NotNull DoubleSupplier throttleInput;
+  private @NotNull DoubleSupplier softTurnInput;
+  private @NotNull DoubleSupplier hardTurnInput;
+
+  /**
+   * Creates a new {@code AdvancedArcadeJoystickInput}.
+   *
+   * @param maxVelocity The maximum velocity of the robot; joystick values will be scaled to this
+   * amount. This should be in position units per second to keep with the specification of {@link
+   * TankDriveData}.
+   * @param trackWidth The track width of the robot (distance between the wheels); this should be in
+   * the same position units as maxVelocity.
+   * @param throttleInput A {@link DoubleSupplier} that supplies the input for the throttle, from -1
```

Comment:
This range doesn't seem to be enforced; should it be, or is not worth the overhead?

Reply:
no me gusta but okay